snet_cli/commands.py

```diff
@@ -813,6 +812,102 @@ def _getserviceregistrationbyname(self, service_json=None):
             self._error(e)
 
+    # Check if service json is valid
+    def _service_json_parser(self, service_json_path):
+        is_ok = True
+        service_json = None
+        try:
+            with open(service_json_path) as f:
+                service_json = json.load(f)
+        except Exception as e:
+            self._printerr("Failed to load {}".format(service_json_path))
+            self._error(e)
+
+        # Get list of valid service_spec/ files
+        valid_import_paths = set()
+        entry_path = None
+        if "service_spec" in service_json and service_json["service_spec"]:
+            spec_path = Path(service_json["service_spec"])
+            if not os.path.isabs(spec_path):
+                entry_path = Path.cwd().joinpath(spec_path).resolve()
+            else:
+                entry_path = spec_path.resolve()
+            if not os.path.isdir(entry_path):
+                self._printerr("Service spec path must resolve to a valid directory: {}".format(spec_path))
+                is_ok = False
+            import_paths = walk_imports(entry_path)
+
+            if import_paths:
+                codegen_tmp_dir = Path("~").expanduser().joinpath(".snet").joinpath("py-codegen").joinpath(
+                    "tmp_publish")
+                if not os.path.exists(codegen_tmp_dir):
+                    os.makedirs(codegen_tmp_dir)
+                proto_include = pkg_resources.resource_filename("grpc_tools", "_proto")
+
+                # Checking if protobuf files are valid by trying to compile them
+                for idx, proto_path in enumerate(import_paths):
+                    protoc_args = [
+                        "protoc",
+                        "-I{}".format(entry_path),
+                        '-I{}'.format(proto_include),
+                        "--python_out={}".format(codegen_tmp_dir),
+                        "--grpc_python_out={}".format(codegen_tmp_dir),
+                        str(proto_path)
+                    ]
+                    # stdout suppression
+                    sys.stdout = open(os.devnull, 'w')
+                    if not protoc(protoc_args):
+                        valid_import_paths.add(proto_path)
+                    else:
+                        sys.stdout = sys.__stdout__
+                        self._printerr("{} is not a valid protobuf file.\n".format(proto_path.name))
+
+                # stdout back to console
+                sys.stdout = sys.__stdout__
+                if os.path.exists(codegen_tmp_dir):
+                    rmtree(codegen_tmp_dir)
+
+        # Create service_spec/ tar and upload it to IPFS
+        ipfs_endpoint = urlparse(self.config["ipfs"]["default_ipfs_endpoint"])
+        ipfs_scheme = ipfs_endpoint.scheme if ipfs_endpoint.scheme else "http"
+        ipfs_port = ipfs_endpoint.port if ipfs_endpoint.port else 5001
+        ipfs_client = ipfsapi.connect(urljoin(ipfs_scheme, ipfs_endpoint.hostname), ipfs_port)
```

Comment:
This IPFS client code is copied three times as far as I see. Could you please wrap it to separate function and reuse.

Reply:
Sure!